quantlib/cashflow.pyx

```diff
@@ -64,69 +65,42 @@ cdef class SimpleCashFlow(CashFlow):
             self.amount, self.date
         )
 
-cdef leg_items(vector[shared_ptr[_cf.CashFlow]] leg):
-    """
-    Returns a list of (amount, pydate)
-    """
-    cdef int i
-    cdef shared_ptr[_cf.CashFlow] _thiscf
-
-    itemlist = []
-    for i in xrange(leg.size()):
-        _thiscf = leg.at(i)
-
-        itemlist.append(
-            (
-                _thiscf.get().amount(),
-                date._pydate_from_qldate(_thiscf.get().date())
-            )
-        )
-    return itemlist
-
 cdef class SimpleLeg:
 
-    def __cinit__(self):
-
-        # Pre-allocated to NULL but intended to be a 
-        # vector[shared_ptr[_cf.CashFlow]]
-        self._thisptr = NULL
-
-    def __init__(self, leg=None, noalloc=False):
-        '''Takes as input a list of (amount, QL Date) tuples. '''
+    def __init__(self, leg=None, bool noalloc=False):
+        '''Takes as input a list of (amount, date) tuples. '''
 
-        #TODO: make so that it handles pydate as well as QL Dates.
         cdef shared_ptr[_cf.CashFlow] _thiscf
-        cdef _date.Date *_thisdate
+        cdef _date.Date _thisdate
 
         if noalloc:
             return
 
-        self._thisptr = new vector[shared_ptr[_cf.CashFlow]]()
-
-        for _amount, _date in leg:
-            _thisdate = (<date.Date>_date)._thisptr.get()
+        for amount, d in leg:
+            if isinstance(d, date.Date):
+               _thisdate = deref((<date.Date>d)._thisptr.get())
+            elif isinstance(d, datetime.date):
+               _thisdate = date._qldate_from_pydate(d)
+            else:
+               raise TypeError("second element needs to be a QuantLib Date or datetime.date")
 
             _thiscf = shared_ptr[_cf.CashFlow](
-                new _cf.SimpleCashFlow(_amount, deref(_thisdate))
+                new _cf.SimpleCashFlow(amount, _thisdate)
             )
 
             self._thisptr.push_back(_thiscf)
 
-    def __dealloc__(self):
```

Comment:
I changed _thisptr into a stack allocated shared_ptr, so no need to clean anything on the python side.